File: qoqo_qasm/tests/acceptance/test_acceptance.py
"""Test qoqo QASM acceptance"""
# Copyright © 2019-2021 HQS Quantum Simulations GmbH. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License
# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
# or implied. See the License for the specific language governing permissions and limitations under
# the License.
import pytest
import sys
import numpy as np
import numpy.testing as npt
from typing import Tuple, Any
from qoqo import operations as ops
from qoqo import Circuit
from qoqo_qasm import qasm_call_operation
from qoqo_qasm import QasmBackend
from qiskit import QuantumCircuit


def test_acceptance_with_qiskit():
    """Test gate operations with QASM interface"""
    circuit = Circuit()

    circuit += ops.RotateX(0, -np.pi)
    circuit += ops.RotateY(0, -np.pi)
    circuit += ops.RotateZ(0, -np.pi)
    circuit += ops.CNOT(0, 1)
    circuit += ops.Hadamard(0)
    circuit += ops.PauliX(0)
    circuit += ops.PauliY(0)
    circuit += ops.PauliZ(0)
    circuit += ops.SGate(0)
    circuit += ops.TGate(0)
    circuit += ops.SqrtPauliX(0)
    circuit += ops.MolmerSorensenXX(0, 1)
    circuit += ops.ControlledPauliY(0, 1)
    circuit += ops.ControlledPauliZ(0, 1)
    circuit += ops.SingleQubitGate(0, 1, 0, 1, 0, 1.0)
    circuit += ops.PragmaRepeatedMeasurement('ro', 1, None)
    circuit += ops.MeasureQubit(0, 'ro', 0)
    circuit += ops.DefinitionFloat(name='rof', length=1, is_output=True)
    circuit += ops.DefinitionBit(name='ro', length=2, is_output=True)
    circuit += ops.DefinitionComplex(name='roc', length=1, is_output=True)
    circuit += ops.InputSymbolic('other', 0)
    circuit += ops.PragmaSetNumberOfMeasurements(20, 'ro')

    backend = QasmBackend(number_qubits=2)
    backend.run_circuit(circuit=circuit, overwrite=True)

    q_circuit = QuantumCircuit.from_qasm_file("default_qasm_backend_output.qasm")


if __name__ == '__main__':
    pytest.main(sys.argv)

File: roqoqo-qasm/tests/acceptance/test_acceptance.rs
// Copyright © 2021-2023 HQS Quantum Simulations GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.
//
//! Test roqoqo QASM acceptance

use qoqo_calculator::CalculatorFloat;
use roqoqo::{operations::*, Circuit};
use roqoqo_qasm::Backend;
use std::env::temp_dir;
use std::fs;
use std::path::Path;

/// Test generating a file for gate operations with QASM interface
#[test]
fn test_acceptance_with_qasmbackend() {
    let backend = Backend::new(Some("qr".to_string()), None);

    let mut circuit = Circuit::new();
    circuit += RotateX::new(0, std::f64::consts::FRAC_PI_2.into());
    circuit += RotateY::new(0, std::f64::consts::FRAC_PI_2.into());
    circuit += RotateZ::new(0, std::f64::consts::FRAC_PI_2.into());
    circuit += CNOT::new(0, 1);
    circuit += Hadamard::new(0);
    circuit += PauliX::new(0);
    circuit += PauliY::new(0);
    circuit += PauliZ::new(0);
    circuit += SGate::new(0);
    circuit += TGate::new(0);
    circuit += SqrtPauliX::new(0);
    circuit += MolmerSorensenXX::new(0, 1);
    circuit += ControlledPauliY::new(0, 1);
    circuit += ControlledPauliZ::new(0, 1);
    circuit += SingleQubitGate::new(
        0,
        CalculatorFloat::from(1),
        CalculatorFloat::from(0),
        CalculatorFloat::from(1),
        CalculatorFloat::from(0),
        CalculatorFloat::from(1.0),
    );
    circuit += PragmaRepeatedMeasurement::new("ro".to_string(), 1, None);
    circuit += MeasureQubit::new(0, "ro".to_string(), 0);
    circuit += DefinitionFloat::new("rof".to_string(), 1, true);
    circuit += DefinitionBit::new("ro".to_string(), 2, true);
    circuit += DefinitionComplex::new("roc".to_string(), 1, true);
    circuit += InputSymbolic::new("other".to_string(), 0.0);
    circuit += PragmaSetNumberOfMeasurements::new(20, "ro".to_string());
    backend
        .circuit_to_qasm_file(
            &circuit,
            temp_dir().as_path(),
            Path::new("test_simple_rust.qasm"),
            true,
        )
        .unwrap();

    let read_in_path = temp_dir().join(Path::new("test_simple_rust.qasm"));
    let b = read_in_path.exists();
    fs::remove_file(&read_in_path).unwrap();
    assert!(b);
}
